### ISSUE TITLE
"error" function being called in Pd calls an "error" function defined in libc.so

### ISSUE THREAD

**Issue by jcelerier** [libpd/libpd#336]
Hello,
I am having a weird case: calls to the "error" function in Pd do not call the "error" defined in s_print.c, but an "error" symbol exported from libc.

e.g here's a screenshot of my debugger in an offending stack trace:

![error](https://user-images.githubusercontent.com/2772730/129452727-9e029ec7-ab8f-4b10-8116-f6cf6fc54fd6.png)

This seems very weird as, the pd symbol is "strongly" exported from my binary (statically linked with libpd):

```
$ nm -A ossia-score | grep -w error
ossia-score:000000000c8a5890 T error
```

and weakly exported from libc:
```
nm -A /usr/lib/libc.so.6 | grep -w error 
/usr/lib/libc.so.6:00000000000fb660 W error
```

thus I don't really know what I could tweak ? I don't quite understand why the linker doesn't prefer the symbol which is part of the executable.

**Comment by jcelerier** [comment]
The error function in libc seems to be that one: https://github.com/bminor/glibc/blob/master/misc/error.c#L270

**Comment by pierreguillot** [comment]
Hi, I think this could be related to this issue https://github.com/pierreguillot/Camomile/issues/239. You can look at these mails https://lists.puredata.info/pipermail/pd-dev/2021-02/022665.html for the fix.

**Comment by jcelerier** [comment]
hi ! yes, using -Bsymbolic would likely fix the issue... but sadly it would break a lot of tools I use during debugging, which work by preloading e.g. replacements to malloc and similar (which Bsymbolic would prevent). I wonder if on the other hand, a version script could be used to mark "error" as symbolic.

**Comment by jcelerier** [comment]
Hmm, after reading https://maskray.me/blog/2021-05-16-elf-interposition-and-bsymbolic
a minimally invasive solution which I found to work, is to make `error` in s_print.c as `__attribute__((visibility("hidden")))`. This way apparently it gets preferred.

**Comment by jcelerier** [comment]
Passing -fvisibility=hidden otoh does not work, likely because the symbol is marked as `extern` in the header ; likely there's an interaction with https://gcc.gnu.org/wiki/Visibility 's following sentence: 

> #pragma GCC visibility is stronger than -fvisibility; it affects extern declarations as well. -fvisibility only affects definitions, so that existing code can be recompiled with minimal changes. This is more true for C than C++; C++ interfaces tend use classes, which are affected by -fvisibility.

**Comment by danomatika** [comment]
The real fix is to change the name of Pd's error function so it doesn't;t conflict with Posix `error()`. This is probably better put as an issue on the main pure-data repo.

**Comment by danomatika** [comment]
Ping @umlaeute @millerpuckette

**Comment by millerpuckette** [comment]
duly pinged :) - I agree, it should be pd_error().  The Pd app itself can also offer error() to maintain binary extern compatibility, but that can be left out of libpd.  I'm in the middle of trying to understand s_audio.c again which will keep me occupied for at leat another week.

**Comment by danomatika** [comment]
The main issue with dropping it I see is using pre-compiled externals which were built with Pd's `error()` from m_pd.h with libpd.

**Comment by millerpuckette** [comment]
And eek, I just remembered there's already a pd_error (that takes a "pd" as first arg).
Question: is binary extern compatibility important for libpd?  If it is then we have to keep error() around.
Also, if _source_ compatibility is important but binary isn't, then I guess m_pd.h should define "error" to something else.

**Comment by umlaeute** [comment]
> A bigger organizational change would be to use `print_` prefix for all the print functions (`print_post()`, `print_error()`, etc...) which aligns with `sys_` etc.

i think a much better organizational change would be to use the **`pd_`** prefix.
we already have `pd_error()` (with slightly different calling conventions, but you really can replace `error(...)` by `pd_error(0, ...)`

**Comment by danomatika** [comment]
Maybe doing a find replace to use pd_error() in the core is a good first step.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com

**Comment by millerpuckette** [comment]
OK.. I think the consensus is... replace all calls to error() in the Pd source
with pd_error(0, ...) and take error() out of m_pd.h but leave it in Pd so that
externs remain binary-compatible... and take it out of libpd altogether.

cheers
Miller

On Mon, Aug 23, 2021 at 05:31:17PM -0700, Dan Wilcox wrote:
> Maybe doing a find replace to use pd_error() in the core is a good first step.
> 
> enohp ym morf tnes
> -----------
> Dan Wilcox
> danomatika.com
> robotcowboy.com
> 
> 
> > On Aug 23, 2021, at 8:24 PM, umläute ***@***.***> wrote:
> > 
> > ﻿
> > A bigger organizational change would be to use print_ prefix for all the print functions (print_post(), print_error(), etc...) which aligns with sys_ etc.
> > 
> > i think a much better organizational change would be to use the pd_ prefix.
> > we already have pd_error() (with slightly different calling conventions, but you really can replace error(...) by pd_error(0, ...)
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> > Triage notifications on the go with GitHub Mobile for iOS or Android.
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_libpd_libpd_issues_336-23issuecomment-2D904227965&d=DwIFaQ&c=-35OiAkTchMrZOngvJPOeA&r=XprZV3Fxus2L1LCw80hE4Q&m=G70Mv37_bW6SN0tqGK210DdUAeed_Xbfy7suNh-4dz8&s=LHAadXLVXFeveFPpoBoztLfYltOm1Yu-eFAMJrrdkLs&e= 

--

**Comment by danomatika** [comment]
This is now fixed in master by using pd 0.52-2.
